src/blog/speeding-up-v8-heap-snapshots.md

```diff
@@ -0,0 +1,161 @@
+
+Some examples of problems with this hash function:
+
+- Once we had strings with a hash key value in lower positions, then the storing of new numbers would offset all the positions.
+- Or even worse: if we would introduce a string with a hash key value that would be a low number, and we already found several hundreds or thousands of consecutive numbers, then that value would be moved several hundreds or thousands of positions.
+
+What did I do to fix it? As the problem comes mostly from numbers represented as strings that would fall in consecutive positions, I modified the hash function so we would rotate the resulting hash value 2 positions to the left. So, for each number, we would introduce 3 free positions. Why 2? Empirical testing across several work-sets showed this number was the best choice to minimize collisions.
+
+[This hashing fix](https://chromium-review.googlesource.com/c/v8/v8/+/4428811) has landed in V8.
+
+## Second Optimization: Caching source positions
+
+After fixing the hashing, we re-profiled and found a further optimization opportunity that would take a significant part of the time. *An old friend*.
+
+For each allocation in the V8 heap, the heap snapshot tries to record the call stack responsible for the allocation. Therefore, for each stack frame, it needs to know the function name and its source location (filename, line number, column number). It was the fetching of this information that turned out to be super slow!
+
+What was happening? It was something similar to [what I fixed in the ETW stack walk](https://chromium-review.googlesource.com/c/v8/v8/+/3669698) and that I explained in [this post](https://blogs.igalia.com/dape/2022/12/21/native-call-stack-profiling-3-3-2022-work-in-v8/).
+
+To compute the source code lines of a function, V8 knows the linear position of the function in the file. But, to find the source line, it needs to traverse the whole file to identify where each newline occurs. This is expensive.
+
+When requesting line information, V8 already knows how to cache the source line positions per-script. But the heap snapshot implementation was not using a cache!
+
+The solution? Before generating the heap snapshot, we now iterate over all the scripts in the V8 context to compute and cache the source line positions.
+
+[This source position caching fix](https://chromium-review.googlesource.com/c/v8/v8/+/4538766) has also landed in V8.
+
+## Did we make it fast?
+
+After enabling both fixes, we re-profiled. Both of our fixes only affect snapshot generation time, so, as expected, snapshot serialization times were unaffected.
+
+When operating on a JS program containing...
+
+- Unoptimized development JS, generation time is **50% faster** 👍
```

Comment:
Maybe `unprocessed`/`processed` would be less ambiguous here? I kept thinking about V8 optimizations when I read `optimized` in this context :)

Reply:
Yes, it is a bit confusing (though it is clarified in next paragraph after this list). Processed/unprocessed looks to me quite ambiguous, and not really improving the readability.

Maybe just removing unoptimized/optimized words could be OK, as we are already talking about development/production JS.

I think I will keep it as is (unless some other reviewer provides an alternate writing that provides a clearly less ambiguous writing?).

Reply:
"development/production JS" seems unambiguous enough to me 👍

Reply:
Yup, I will change writing to make that less ambiguous, providing a better explanation of how production source code is optimized.